pattern/variable/Variable.java

```diff
@@ -71,6 +78,34 @@ public ThingVariable asThing() {
         throw GraknException.of(INVALID_CASTING, className(this.getClass()), className(ThingVariable.class));
     }
 
+    public void addHints(Set<Label> labels) {
+        if (typeHints.isEmpty() && isSatisfiable) return;
+        typeHints.addAll(labels);
+        if (!isSatisfiable && !typeHints.isEmpty()) isSatisfiable = true;
+    }
+
+    public void retainHints(Set<Label> labels) {
+        if (typeHints.isEmpty() && isSatisfiable) {
+            typeHints.addAll(labels);
+            return;
+        }
+        typeHints.retainAll(labels);
+        if (typeHints.isEmpty()) isSatisfiable = false;
+    }
+
+    public void removeHint(Label label) {
+        assert !(typeHints.isEmpty() && isSatisfiable);
+        typeHints.remove(label);
+    }
```

Comment:
removeHint() is only used when removing types that don't conform to being a sub of a hint of the variables sub constraint. When the Variable has no typeHints, this method is not used, and instead all the subTypes of it's sub constraint is inserted as its TypeHints.